arbor/backends/gpu/matrix_state_fine.hpp

```diff
@@ -300,53 +312,63 @@ struct matrix_state_fine {
         // kernel.
-        levels.reserve(total_num_levels);
-        levels_start.reserve(branch_maps.size() + 1);
-        levels_start.push_back(0);
-        data_partition.reserve(branch_maps.size());
-        // offset into the packed data format, used to apply permutation on data
+
+        // Accumulate metadata about the levels, level lengths, level parents,
+        // data_partition and block indices in temporary vectors to be copied to the device
+        std::vector<level_metadata> temp_meta;
+        std::vector<size_type> temp_lengths, temp_parents, temp_data_part, temp_block_index;
+
+        temp_block_index.reserve(branch_maps.size() + 1);
+        temp_block_index.push_back(0);
+        temp_data_part.reserve(branch_maps.size());
+
+        // Offset into the packed data format, used to apply permutation on data
         auto pos = 0u;
+        // Offset into the packed data format, used to access levels_lengths and levels_parents
+        auto data_start = 0u;
         for (const auto& branch_map: branch_maps) {
             for (const auto& lvl_branches: branch_map) {
 
-                level lvl(lvl_branches.size());
+                level_metadata lvl_meta;
+                std::vector<size_type> lvl_lengths(lvl_branches.size()), lvl_parents(lvl_branches.size());
+
+                lvl_meta.num_branches = lvl_branches.size();
+                lvl_meta.matrix_data_index = pos;
+                lvl_meta.level_data_index = data_start;
 
                 // The length of the first branch is the upper bound on branch
                 // length as they are sorted in descending order of length.
-                lvl.max_length = lvl_branches.front().length;
-                lvl.data_index = pos;
+                lvl_meta.max_length = lvl_branches.front().length;
 
                 unsigned bi = 0u;
                 for (const auto& b: lvl_branches) {
                     // Set the length of the branch.
-                    lvl.lengths[bi] = b.length;
+                    lvl_lengths[bi] = b.length;
 
                     // Set the parent indexes. During the forward and backward
                     // substitution phases each branch accesses the last node in
                     // its parent branch.
                     auto index = b.parent_id==npos? npos: branch_locs[b.parent_id].index;
-                    lvl.parents[bi] = index;
+                    lvl_parents[bi] = index;
                     ++bi;
                 }
 
-                pos += lvl.max_length*lvl.num_branches;
+                data_start+= lvl_meta.num_branches;
+                pos += lvl_meta.max_length*lvl_meta.num_branches;
 
-                levels.push_back(std::move(lvl));
+                temp_meta.push_back(std::move(lvl_meta));
+                std::move(lvl_lengths.begin(), lvl_lengths.end(), std::back_inserter(temp_lengths));
+                std::move(lvl_parents.begin(), lvl_parents.end(), std::back_inserter(temp_parents));
```

Comment:
I didn't know that `std::move` was also an algorithm! Because we are copying integers, `std::copy` would be just as efficient.

Or you could use `util::append(temp_parents, lvl_parents)` from `util/rangeutil.hpp`.

Reply:
I'll use `util::append`, thanks!